File: Submissions/609_Find_Duplicate_File_in_System.cpp
class Solution {
public:
    vector<vector<string>> findDuplicate(vector<string>& paths) {
        vector<vector<string> > result;
        map<string, vector<string> > m;
        for(auto path : paths) {
            int pos = path.find(' ');
            string directory = path.substr(0, pos);
            ++pos;
            while(pos < path.length()) {
                int next_pos = path.find('(', pos);
                string filename = path.substr(pos, next_pos-pos);
                pos = next_pos;
                
                next_pos = path.find(')', pos);
                string content = path.substr(pos, next_pos-pos);
                pos = next_pos + 2;
               
                m[content].push_back(directory + "/" + filename);
            }
        }
        
        for(auto iter=m.begin(); iter!=m.end(); ++iter) {
            if(iter->second.size() > 1) result.push_back(iter->second);
        }
        
        return result;
    }
};

File: Kotlin/609_Find_Duplicate_File_in_System.kt
class Solution {
    fun findDuplicate(paths: Array<String>): List<List<String>> {
        var result = ArrayList<List<String>>()
        val m = HashMap<String, HashSet<String>>()
        
        for(path in paths){
            val list = path.split(' ')
            val dir = list[0]
            for(i in 1 until list.size){
                val pos = list[i].indexOf('(')
                val next_pos = list[i].indexOf(')')
                val file = list[i].substring(0, pos)
                val content = list[i].substring(pos + 1, next_pos)
                val s = m.getOrDefault(content, hashSetOf())
                s.add(dir + "/" + file)
                m.put(content, s)
            }
        }
        
        m.forEach{
            if (it.value.size > 1) result.add(it.value.toList())
        }
        return result
    }
}